src/flint/fmpz_factor.jl

```diff
@@ -18,30 +18,45 @@ function _factor(a::ZZRingElem)
 end
 
 # Just to give a helpful error message if someone tries to factor a boolean
-function factor(b::Bool)
-  throw(DomainError("Cannot factorize a boolean"));
+function factor(::Bool)
+  throw(DomainError("Cannot factorize a boolean"))
 end
 
-# This function handles machine integer types (up to 64 bits)
-function factor(a::T) where T <: Union{Int, UInt}
+function factor(::Type{Bool}, ::Int)
+  throw(DomainError("Cannot have a factorization into booleans"))
+end
+
+function factor(::Type{Bool}, ::ZZRingElem)
+  throw(DomainError("Cannot have a factorization into booleans"))
+end
+
+
+# The main dispatch function: flintify gives either Int or ZZRingElem,
+# then use Julia's dispatcher to select the worker function.
+function factor(a::T)  where T <: Integer
+  iszero(a) && throw(ArgumentError("Argument must be non-zero"))
```

Comment:
```suggestion
  @req !iszero(a) "Argument must be non-zero"
```

Reply:
Ah!  This changes the error type from `DomainError` (which seems most suitable) to `MethodError` (which seems rather vague).  I definitely prefer `DomainError`; do you prefer `MethodError`?

Reply:
Huh? Your current code throws a `ArgumentError` (not a `DomainError`). And `@req` also throws a `ArgumentError` (not a `MethodError`).

Reply:
Sorry, my  mistake: the problem was the ambiguity mentioned in the previous comment block.